Log Prob Visualiser

neelnanda-io opened the pull request:
Added a log prob visualizer! This shows you some text, with tokens coloured by the log prob of the model. If you hover over a token, it shows you the top 10 tokens predicted, along with logits and log probs, and the values + rank of the correct token. The correct token is highlighted in the list of top 10 if it occurs in there (if so, it appears twice).

![image](https://user-images.githubusercontent.com/77788841/209418803-ac23ed13-fed3-47fe-b57c-d4d7bd092715.png)

![image](https://user-images.githubusercontent.com/77788841/209418807-1cb95f00-757c-43c6-8184-3b26550716fc.png)

neelnanda-io commented:
Did some internal refactoring of the Token class - pulled out the formatting text part into a separate function, and added a `TokenCustomTooltip` class that can display an arbitrary React Component as its tooltip.

I also have some local code that automates the creation of BS boilerplate (for writing the visualization and for mocking)

alan-cooney reviewed (changes requested):
A few small changes in the comments - note the checks are also failing so there may be an issue still after these are made

neelnanda-io commented on react/src/repl_tf.js:
```diff
@@ -0,0 +1,36 @@
+const tf = require('@tensorflow/tfjs');
+const readline = require('readline');
+
+const React = require("react");
+const { Container, Row, Col } = require("react-grid-system");
+// const col = require("./logits/LogProbVis")
+// const ColoredTokens = require("./tokens/ColoredTokens").ColoredTokens;
+// const TokenCustomTooltip = require("./tokens/utils/TokenCustomTooltip").TokenCustomTooltip;
+// const Token = require("./tokens/utils/Token").Token;
+// const formatTokenText = require("./tokens/utils/Token").formatTokenText;
+
+const colord = require("colord");
+// const extend = require("colord/lib/extend");
+// const AnyColor = require("colord/lib/types/AnyColor");
+const Colord = colord.Colord
+const mixPlugin = require("colord/plugins/mix");
+const namesPlugin = require("colord/plugins/names");
+colord.extend([mixPlugin, namesPlugin]);
+
+const rl = readline.createInterface({
+    input: process.stdin,
+    output: process.stdout
+});
+
+console.log('Welcome to the TensorFlow.js REPL!');
+console.log('Enter a TensorFlow.js expression or statement and press enter to evaluate it.');
+
+rl.on('line', (input) => {
+    try {
```
Oh lol, nope.

neelnanda-io commented:
Not sure how the dependency changes happened!

neelnanda-io commented:
All changes made

neelnanda-io commented:
I did not notice that there were tests! Turns out I had an actual bug, so good shout (imported some things twice in index.ts), fixed.
Also funky, my local code thought the library was v 1.6.0 not 1.34.0, which broke the topk test. I added a thing which monkey patches the version to avoid each update breaking this, but happy to revert if you want. It _did_ catch a real bug in my case, but will also break for each new version.
